### PR TITLE
chore(databases-collection): refactor create db / create coll to a single create namespace plugin COMPASS-7410

### PR DESCRIPTION
**Pull request by gribnoysup**
Last follow-up to https://github.com/mongodb-js/compass/pull/5085

- Merges create database and create collection plugin into a single create namespace plugin
- Store is updated to typescript, but all existing logic is preserved with minimal changes
- Plugin is using service injectors for everything, including the new instance locator

### CODE REVIEW

File: packages/compass-app-stores/src/provider.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { createContext, useContext } from 'react';
 
 export const InstanceContext = createContext<MongoDBInstance | null>(null);
 
-export const useMongoDBInstance = (): MongoDBInstance => {
+export const mongoDBInstanceLocator = (): MongoDBInstance => {
```
Comment (gribnoysup) [inline packages/compass-app-stores/src/provider.ts:6]: @addaleax don't know if you have a strong preference here, I thought that I'd rename it to match other methods we have and to indicate a bit better that it's not supposed to be used as a hook in React (even though it sorta is).

I think if we want to add actual hooks here (should be very valuable in some cases!), those should probably be a bit more high level, encapsulating the model subscription logic for a required field internally instead of just exposing the modal to the React view 🤔

Comment (addaleax) [reply packages/compass-app-stores/src/provider.ts:6]: sure, no strong preferences here! happy to stick to whatever pattern we choose :slightly_smiling_face:

File: packages/instance-model/index.d.ts [hunk 2]
```diff
@@ -100,6 +100,7 @@ declare class MongoDBInstance extends MongoDBInstanceProps {
   }): Promise<Collection | null>;
   removeAllListeners(): void;
   on(evt: string, fn: (...args: any) => void);
+  off(evt: string, fn: (...args: any) => void);
```
Comment (addaleax) [inline packages/instance-model/index.d.ts:103]: lol :smile: https://github.com/mongodb-js/compass/pull/5102/files#diff-a79e9b0455facd1015dd43ba2327be7c140c049fcbb16ef624e15a67dfc38cbaR103 no strong feelings either way will align that with this PR ;)

Comment (addaleax) [reply packages/instance-model/index.d.ts:103]: ah actually, now I remember why I did this; `AppRegistry` has `removeListener` but not `off`. Should we also add `off` to `AppRegistry` then?

Comment (gribnoysup) [reply packages/instance-model/index.d.ts:103]: We can do that for consistency, yes, especially now that we use it so much! With ampersand models you have off, but not removeListener, so this is why I added it here

**Comment by gribnoysup** [comment]
Added tests, should be ready for review now

File: packages/hadron-app-registry/src/register-plugin.tsx [hunk 3]
```diff
@@ -78,8 +88,29 @@ export type HadronPluginConfig<T, S extends Record<string, () => unknown>> = {
...
+   *
+   * @param mocks Overrides for the services locator values and registries
+   *              passed to the plugin in runtime. When `globalAppRegistry`
+   *              override is passed, it will be also used for the
+   *              localAppRegistry override unless `localAppRegistry` is also
+   *              explicitly passed as a mock service.
+   */
+  withMockServices(
+    mocks: Partial<Registries & Services<S>>
+  ): React.FunctionComponent<T>;
```
Comment (gribnoysup) [inline packages/hadron-app-registry/src/register-plugin.tsx:113]: @addaleax this is this new testing helper I mentioned, I also switched drop-namespace tests to use it too (although due to how this particular plugin works, it doesn't make that much of a difference in that case)

File: packages/hadron-app-registry/src/register-plugin.tsx [hunk 4]
```diff
@@ -194,6 +225,36 @@ export function registerHadronPlugin<
...
+          globalAppRegistry: _globalAppRegistry,
+          localAppRegistry: _localAppRegistry,
+          ...mockServices
+        } = mocks;
+        const mockServiceLocators = Object.fromEntries(
+          Object.keys(services ?? {}).map((s) => {
+            return [s, mockServices[s] ? () => mockServices[s] : services?.[s]];
+          })
+        ) as unknown as S;
+        const MockPlugin = registerHadronPlugin(config, mockServiceLocators);
```
Comment (addaleax) [inline packages/hadron-app-registry/src/register-plugin.tsx:241]: Won’t it be an issue that this re-uses the same `registryName` as the original plugin and therefore shares a store (which may have already been initialized with the original services)?

Comment (gribnoysup) [reply packages/hadron-app-registry/src/register-plugin.tsx:241]: Hmmm, I think for that to conflict you'd need to actually render / activate both mocked and unmocked plugin in the same test, which sounds like a rare corner case, unless I'm missing some scenario here 🤔 Doesn't hurt to  override config name here though, generally not that we should rely on it in any way externally, can do that!

Comment (addaleax) [reply packages/hadron-app-registry/src/register-plugin.tsx:241]: Just needs to happen in the same process as long as the global app registry is truly global, right? That doesn’t seem all _that_ unlikely, and it’s going to be hard to diagnose because it could involve some “spooky action at a distance” from a completely unrelated test that also happens to render the plugin.

Another approach, maybe even a better one, would be figuring out how to best avoid having a truly global globalAppRegistry in tests at all. Ideally, tests should probably use a distinct one each time, not share the same on across all tests in that process…?

Comment (gribnoysup) [reply packages/hadron-app-registry/src/register-plugin.tsx:241]: Ah, I see what you mean. Then I think we can change the code there that passes `globalAppRegistry` value to the context to be `_globalAppRegistry ?? new AppRegistry()`: if you don't care about the reigstry or don't need to interact with it inside the actual test, this would be just a safe fallback, but if you want to emit events or something, then you'd want to pass your instance there. How does that sound?

Comment (addaleax) [reply packages/hadron-app-registry/src/register-plugin.tsx:241]: Yeah, I think that makes for some nice isolation between tests, and it should be easy enough to work around if it we do want the actual global instance?

Comment (gribnoysup) [reply packages/hadron-app-registry/src/register-plugin.tsx:241]: Totally, and it would be super explicit if you want to pass the global one, it's a singleton export from the app-registry package. Will update, thanks for the feedback!

File: packages/instance-model/index.d.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/instance-model/index.d.ts]: Renamed to align with your PR in https://github.com/mongodb-js/compass/pull/5100/commits/6691588aff5e4ba93031a41042309b9be3b182e0

File: packages/hadron-app-registry/src/register-plugin.tsx (hunk 4 above)
Comment (gribnoysup) [reply packages/hadron-app-registry/src/register-plugin.tsx:241]: Updated `globalAppRegistry` to be always scoped as discussed in https://github.com/mongodb-js/compass/pull/5100/commits/72c9c1587fe316133f4b3c43e1c20e5665dbd946. Also when doing that stumled on something that I think is a bug: we did call `plugin.deactivate` in `appRegistry.deactivate`, but didn't de-register plugins, changed that too to always clean-up plugins from registry as at the moment we call deactivate there, we expect this whole thing to be disposeable basically

**Review by addaleax — APPROVED** [review]
